doc/design/manifest.md

```diff
@@ -0,0 +1,589 @@
+
+We do not support channels or groupings of components. We expect components will support something here, but it will be up to the client to implement.
+
+TiUp itself can be treated as a component and I think does not need to be treated any differently on the server side.
+
+## Other terms
+
+* repository: a source of components and manifests (aka server, mirror)
+* manifest: a JSON metadata file
+* platform: the hardware platform and operating system where a binary will run (c.f., built).
+
+
+## Principles
+
+* URLs are persistent.
+* All data is immutable
+* Most metadata is immutable, achieved using versioning.
+* Components cannot be removed or deleted.
+* Zero-trust access control.
+* Minimise the number of keys and the frequency with which they must be replaced.
+* Uploads and downloads should be secure.
+* All operations should be idempotent or at least repeatable. If an operation is interupted, the user can just retry until it succeeds.
+* There should be no 'transaction' mechanism between client and server, groups of operations should not require atomicity.
+* Minimise required knowledge in the client - the metadata should direct the client how to perform operations with minimal knowledge of the server's data layout, etc.
+* We should not require consistency between files in the repo (beyond hashes and keys).
+
+
+## Assumptions
+
+* Updates are per-component, updating all components is interpreted by the client.
+* There is a secure channel to distribute initial and updated keys for each server
+* Don't need to change the owner of a component.
+* Don't need to delegate ownership of a component (owners can delegate themselves with client-side tooling).
+
+
+## Repository organisation
+
+We require snapshot.json and root.json to exist at the root of the repository. There will be many n.root.json files, these must also be at the repository root.
+
+The rest of the repository organisation is here for completeness but can be changed freely since all other locations are specified by URLs.
+
+n.index.json files and snapshot.json are at the root of the repository.
+
+For each component there is a directory named with the component's id. Within this directory are the n.id.json files (where id is the id of the component; alternative: all these manifests could be called n.component.json). Within the component directory, there is a sub-directory for each platform named with the target triple. All data files are within these platform subdirectories. Each file has the form version.platform.id.tar.gz for tarballs and version.platform.id.ext or version.platform.id for other single files, where id is the component id, platform is a target triple, and version is a semver version containing exactly two periods (`.`, e.g., `1.0.0`).
```

Comment:
The version-first naming scheme seems very strange. Id-first provides a more natural sorted order. I can't think of an advantage to version-first. `id.version.platform` makes more sense to me.

The subdirectory-per platform is redundant with the file-naming scheme. That's not necessarily a problem, since it could e.g. aid human comprehension.

Some rationale for both these choices would help me.

Reply:
Perhaps the version-first scheme is related to ease of parsing?

Reply:
There are conceivably platformless components too. Consider how a component with no target-triple would fit into this scheme.

Reply:
> The version-first naming scheme seems very strange. Id-first provides a more natural sorted order. I can't think of an advantage to version-first. id.version.platform makes more sense to me.

Good point, there was an older version where it was important for the version to be first but now it doesn't matter and the ordering is good.

> The subdirectory-per platform is redundant with the file-naming scheme. That's not necessarily a problem, since it could e.g. aid human comprehension.

Yep, rationale is purely to make it easier for manual administration, but it sounds like using just one directory is better for the CDN.